### PR TITLE
[runtime] Switch mono-hash implementation from chaining to open adressing

### PR DESCRIPTION
**Pull request by ludovic-henry**
This improves CPU cache utilisation, which is very important for the GC marking phase.

### CODE REVIEW

**Comment by schani** [comment]
builddeb

File: mono/metadata/mono-hash.c [hunk 1]
```diff
@@ -43,12 +44,16 @@
...
+
 typedef struct _Slot Slot;
 
 struct _Slot {
 	MonoObject *key;
 	MonoObject *value;
-	Slot    *next;
+	gsize is_used : 1;
+	gsize is_tombstone : 1;
+	gsize distance : sizeof (gsize) - 2;
```
Comment (kumpera) [inline mono/metadata/mono-hash.c]: This will be 2 bits for 32 bits and 6 bits on 64 bits.

**Comment by lewurm** [comment]
`<botmode>`
Benchmark results: http://xamarin.github.io/benchmarker/front-end/pullrequest.html#id=22
buildbot logs: https://performancebot.mono-project.com/builders/debian-amd64_benchmarker_pullrequest_auto-sgen-noturbo/builds/24
`</botmode>`

**Comment by schani** [comment]
Looks like it's making F# faster, IronPython slower.  Here is a comparison with commits before and after the pull request base commit, just to make sure the differences aren't due to the one commit (bumping reference sources) in between:

  http://xamarin.github.io/benchmarker/front-end/compare.html#ids=1098+1101+1103

**Comment by ludovic-henry** [comment]
@schani I will work on implementing http://codecapsule.com/2013/11/17/robin-hood-hashing-backward-shift-deletion/
That should drastically reduce collisions and thus big Distance to Initial Bucket

**Comment by ludovic-henry** [comment]
Compiled corlib with Valgrind without -O2 optimization:

```
Without open hash addressing

==62733==
==62733== Events    : Ir Dr Dw I1mr D1mr D1mw ILmr DLmr DLmw AcCost1 SpLoss1 AcCost2 SpLoss2
==62733== Collected : 25985976113 8797254112 4839927538 169506822 116349224 31986298 1204172 23606387 13820403 77604006067 10264833554 6679212415 704148864
==62733==
==62733== I   refs:      25,985,976,113
==62733== I1  misses:       169,506,822
==62733== LLi misses:         1,204,172
==62733== I1  miss rate:           0.65%
==62733== LLi miss rate:           0.00%
==62733==
==62733== D   refs:      13,637,181,650  (8,797,254,112 rd + 4,839,927,538 wr)
==62733== D1  misses:       148,335,522  (  116,349,224 rd +    31,986,298 wr)
==62733== LLd misses:        37,426,790  (   23,606,387 rd +    13,820,403 wr)
==62733== D1  miss rate:            1.1% (          1.3%   +           0.7%  )
==62733== LLd miss rate:            0.3% (          0.3%   +           0.3%  )
==62733==
==62733== LL refs:          317,842,344  (  285,856,046 rd +    31,986,298 wr)
==62733== LL misses:         38,630,962  (   24,810,559 rd +    13,820,403 wr)
==62733== LL miss rate:             0.1% (          0.1%   +           0.3%  )

With open hash addressing

==71144==
==71144== Events    : Ir Dr Dw I1mr D1mr D1mw ILmr DLmr DLmw AcCost1 SpLoss1 AcCost2 SpLoss2
==71144== Collected : 26031051227 8812074582 4846612692 169694377 116979618 32193802 1196324 23541367 13753330 78090398985 10309747874 6623889349 698988576
==71144==
==71144== I   refs:      26,031,051,227
==71144== I1  misses:       169,694,377
==71144== LLi misses:         1,196,324
==71144== I1  miss rate:           0.65%
==71144== LLi miss rate:           0.00%
==71144==
==71144== D   refs:      13,658,687,274  (8,812,074,582 rd + 4,846,612,692 wr)
==71144== D1  misses:       149,173,420  (  116,979,618 rd +    32,193,802 wr)
==71144== LLd misses:        37,294,697  (   23,541,367 rd +    13,753,330 wr)
==71144== D1  miss rate:            1.1% (          1.3%   +           0.7%  )
==71144== LLd miss rate:            0.3% (          0.3%   +           0.3%  )
==71144==
==71144== LL refs:          318,867,797  (  286,673,995 rd +    32,193,802 wr)
==71144== LL misses:         38,491,021  (   24,737,691 rd +    13,753,330 wr)
==71144== LL miss rate:             0.1% (          0.1%   +           0.3%  )
```

**Comment by schani** [comment]
I don't think this should be merged until the full thing is implemented and performance evaluated.

**Comment by ludovic-henry** [comment]
I agree with @schani

**Pull request closed without merging**